[GUI] Implement Stackable Text Outline on `Label`

Delsin-Yu opened the pull request:
Closes: godotengine/godot-proposals#2564

This PR implements `Stacked Outlines` and `Stacked Shadows` for the `Label` type.
Engine users may configure the effect through `LabelSettings`.

![image](https://github.com/user-attachments/assets/4d41ee6d-49f8-4522-9e4b-674bb1006187)

https://github.com/user-attachments/assets/31b2d724-b3f0-431a-a413-985b2079af31

Delsin-Yu commented:
Labels in Godot always have their drop shadow renders behind the outlines. However, this image from the proposal raises the question: What do we do with the stackable effects?

Also, the way this feature is implemented involves two additions to the current theme system. Are there better alternatives?

![image](https://github.com/user-attachments/assets/486d3d54-3eb7-49ce-8375-ee0062efd85e)

Delsin-Yu commented:
I did not handle the new enum types in `doc_tools`, but the current commit should compile if using simple scons command (`scons dev_build=yes`).

Delsin-Yu commented:
This [specific commit](https://github.com/godotengine/godot/pull/104731/commits/0fe221022e1abf293435de610d8083f7a047de63) can be controversial, as flattens render loop (for better maintenance clarity).

The reason is that too much branch logic inside one single text line drawing loop complicates the ongoing implementation of stacked drop shadow (and its outline) drawings.

It looks like clang doesn't like zero length `__VAR_ARGS__`, I will commit a fix shortly.

Delsin-Yu commented:
> Will gradient outlines be implemented in this pr?

As `Label` does not initially support gradients, I think No. That should be implemented in `RichTextLabel`. Let's see if anyone with better expertise can offer a follow-up PR implementing Gradient-related features in `RichTextLabel` (the source code in RTL is a little too complex for me to read through, but I can try if no one is available).

Delsin-Yu commented:
While the stacked text shadow is capable of covering the function of stacked outline, stacked outline provides better usability, as the outline size is calculated automatically for them, for example, if the user sets outline size to be [3, 3], the actual outline size will be [3, 6], which is more intuitive.

Calinou reviewed (commented):
Tested locally, it works as expected. I've also tested (re)setting color/constant arrays on multiple selected nodes at once and it works too. It also works with MSDF font rendering, although remember to increase the MSDF pixel range for high outline sizes in the Import dock as usual.

**Testing project:** [test_pr_104731.zip](https://github.com/user-attachments/files/19539210/test_pr_104731.zip)

![image](https://github.com/user-attachments/assets/c9b3bd66-0a75-4358-ba7c-7ccbce81b02d)


Some feedback:

- When you first set a theme color/constant array override, it changes from Nil to an array of size 0, which looks a bit strange:

![image](https://github.com/user-attachments/assets/7c635246-ff42-4ee4-87b2-a34bdcb3e4a2)

I think non-overridden items should be empty arrays instead, so that their value doesn't change as soon as you add an override. This is also more consistent with non-array constants which start at `0`.

- There should be a node configuration warning when there's a mismatch between the number of outline/shadow colors and sizes/offsets, as values that don't have a matching will fall back to the default (transparent for color, `0` for size/offset), which is probably not intended by the user.
- Changes to colors within constant arrays are not applied instantly, unlike change to non-array theme colors. You need to click outside the color picker for the change to apply, which slows down iteration.
  - This may be a more general issue with the array property editors, I'm not sure.

Delsin-Yu commented:
> I think non-overridden items should be empty arrays instead

I will have to look into it, but I think it's Godot's default behavior if I don't implement any specific logic (I merely imitate other implementations).

> There should be a node configuration warning when there's a mismatch between the number of outline/shadow colors and sizes/offsets

Yes, I will try to implement that.

> Changes to colors within constant arrays are not applied instantly ... This may be a more general issue with the array property editors

I will see what I can do about that, but I haven't looked at `EditorPropertyArray` yet. I ended up implementing a dedicated array property inspector for the theme editor.

bruvzg commented on scene/gui/label.cpp:
```diff
@@ -430,20 +434,75 @@ inline void draw_glyph_shadow(const Glyph &p_gl, const RID &p_canvas, const Colo
 	}
 }
 
+#define DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para_start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, color, draw_func, ...) \
```
Using `inline` function probably would be more readable.

Delsin-Yu replied:
I'm worried about performance penalties for inline causes, as the compiler does not guarantee the function will be inlined?

bruvzg replied:
It's not guaranteed, but should not be a performance issue.

Delsin-Yu replied:
Okay, I will change that back to the inline function call.

Delsin-Yu replied:
I'm having trouble converting the MACRO to an inline function; the main blocker is passing vararg parameters pack into the function pointer that contains vararg:

![dfc844f6c00e7be7e3f5fc8bd419bf13](https://github.com/user-attachments/assets/1e16cbec-830e-4549-907a-e0193e472da3)

![c64b91a10a52486e400a5ea9b32f54a6](https://github.com/user-attachments/assets/4ea79bcb-cfb6-42cb-b650-068ccd7f355b)

Changing the argument type from `Vector2` to `Point2` in `draw_glyph_shadow_ouline` also won't work because of the const annotation.

![a9f3933c75e7239ebf6f30460f5b6bb6](https://github.com/user-attachments/assets/3a8251c0-1a7f-4cee-8cb0-4e9524177cbd)

bruvzg reviewed (commented):
I would consider only using `LabelSettings` instead of theme. New properties can be also bound in the similar way `items` work in the `PopupMenu` and `OptionButton` to make a more convenient editor inspector UI.

AThousandShips commented:
Are you using some plugin to suggest includes? If so please turn it off it suggests a lot of incorrect cases

Delsin-Yu commented:
> If so please turn it off it suggests a lot of incorrect cases

Oh, it was old includes from the previous design approach, I forget to removed them.

Delsin-Yu committed:
Implmement stackable text effects on label type through label settings

Co-authored-by: Micky <66727710+Mickeon@users.noreply.github.com>
Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

Delsin-Yu commented:
I have removed all modifications to the `Theme` and implemented the stacked effects storage only in `LabelSettings`, this also enables better UI/UX.

![image](https://github.com/user-attachments/assets/bc202e59-fa27-4da9-8b6b-e1cca2c35abe)

beicause commented:
Here is a patch that addresses https://github.com/godotengine/godot/pull/104731#discussion_r2024817685 

```diff
diff --git a/scene/gui/label.cpp b/scene/gui/label.cpp
index 4387577e58..ccd861981a 100644
--- a/scene/gui/label.cpp
+++ b/scene/gui/label.cpp
@@ -445,61 +445,6 @@ inline void draw_glyph_outline(const Glyph &p_gl, const RID &p_canvas, const Col
 	}
 }
 
-#define DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para_start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, color, draw_func, ...) \
-	do {                                                                                                                                                                                                                                                                           \
-		processed_glyphs_step = processed_glyphs;                                                                                                                                                                                                                                  \
-		Vector2 offset_step = ofs;                                                                                                                                                                                                                                                 \
-		/* Draw RTL ellipsis string when necessary. */                                                                                                                                                                                                                             \
-		if (rtl && ellipsis_pos >= 0) {                                                                                                                                                                                                                                            \
-			for (int gl_idx = ellipsis_gl_size - 1; gl_idx >= 0; gl_idx--) {                                                                                                                                                                                                       \
-				for (int j = 0; j < ellipsis_glyphs[gl_idx].repeat; j++) {                                                                                                                                                                                                         \
-					bool skip = (trim_chars && ellipsis_glyphs[gl_idx].end + para_start > visible_chars) || (trim_glyphs_ltr && (processed_glyphs_step >= visible_glyphs)) || (trim_glyphs_rtl && (processed_glyphs_step < total_glyphs - visible_glyphs));                        \
-					if (!skip) {                                                                                                                                                                                                                                                   \
-						draw_func(ellipsis_glyphs[gl_idx], ci, color, offset_step, ##__VA_ARGS__);                                                                                                                                                                                 \
-					}                                                                                                                                                                                                                                                              \
-					processed_glyphs_step++;                                                                                                                                                                                                                                       \
-					offset_step.x += ellipsis_glyphs[gl_idx].advance;                                                                                                                                                                                                              \
-				}                                                                                                                                                                                                                                                                  \
-			}                                                                                                                                                                                                                                                                      \
-		}                                                                                                                                                                                                                                                                          \
-		/* Draw main text. */                                                                                                                                                                                                                                                      \
-		for (int j = 0; j < gl_size; j++) {                                                                                                                                                                                                                                        \
-			/* Trim when necessary. */                                                                                                                                                                                                                                             \
-			if (trim_pos >= 0) {                                                                                                                                                                                                                                                   \
-				if (rtl) {                                                                                                                                                                                                                                                         \
-					if (j < trim_pos) {                                                                                                                                                                                                                                            \
-						continue;                                                                                                                                                                                                                                                  \
-					}                                                                                                                                                                                                                                                              \
-				} else {                                                                                                                                                                                                                                                           \
-					if (j >= trim_pos) {                                                                                                                                                                                                                                           \
-						break;                                                                                                                                                                                                                                                     \
-					}                                                                                                                                                                                                                                                              \
-				}                                                                                                                                                                                                                                                                  \
-			}                                                                                                                                                                                                                                                                      \
-			for (int k = 0; k < glyphs[j].repeat; k++) {                                                                                                                                                                                                                           \
-				bool skip = (trim_chars && glyphs[j].end + para_start > visible_chars) || (trim_glyphs_ltr && (processed_glyphs_step >= visible_glyphs)) || (trim_glyphs_rtl && (processed_glyphs_step < total_glyphs - visible_glyphs));                                          \
-				if (!skip) {                                                                                                                                                                                                                                                       \
-					draw_func(glyphs[j], ci, color, offset_step, ##__VA_ARGS__);                                                                                                                                                                                                   \
-				}                                                                                                                                                                                                                                                                  \
-				processed_glyphs_step++;                                                                                                                                                                                                                                           \
-				offset_step.x += glyphs[j].advance;                                                                                                                                                                                                                                \
-			}                                                                                                                                                                                                                                                                      \
-		}                                                                                                                                                                                                                                                                          \
-		/* Draw LTR ellipsis string when necessary. */                                                                                                                                                                                                                             \
-		if (!rtl && ellipsis_pos >= 0) {                                                                                                                                                                                                                                           \
-			for (int gl_idx = 0; gl_idx < ellipsis_gl_size; gl_idx++) {                                                                                                                                                                                                            \
-				for (int j = 0; j < ellipsis_glyphs[gl_idx].repeat; j++) {                                                                                                                                                                                                         \
-					bool skip = (trim_chars && ellipsis_glyphs[gl_idx].end + para_start > visible_chars) || (trim_glyphs_ltr && (processed_glyphs_step >= visible_glyphs)) || (trim_glyphs_rtl && (processed_glyphs_step < total_glyphs - visible_glyphs));                        \
-					if (!skip) {                                                                                                                                                                                                                                                   \
-						draw_func(ellipsis_glyphs[gl_idx], ci, color, offset_step, ##__VA_ARGS__);                                                                                                                                                                                 \
-					}                                                                                                                                                                                                                                                              \
-					processed_glyphs_step++;                                                                                                                                                                                                                                       \
-					offset_step.x += ellipsis_glyphs[gl_idx].advance;                                                                                                                                                                                                              \
-				}                                                                                                                                                                                                                                                                  \
-			}                                                                                                                                                                                                                                                                      \
-		}                                                                                                                                                                                                                                                                          \
-	} while (false)
-
 void Label::_ensure_shaped() const {
 	if (dirty || font_dirty || text_dirty) {
 		_shape();
@@ -808,7 +753,7 @@ void Label::_notification(int p_what) {
 			int font_h = font->get_height(font_size);
 			Color font_color = has_settings ? settings->get_font_color() : theme_cache.font_color;
 			Color font_shadow_color = has_settings ? settings->get_shadow_color() : theme_cache.font_shadow_color;
-			Point2 shadow_ofs = has_settings ? settings->get_shadow_offset() : theme_cache.font_shadow_offset;
+			const Point2 shadow_ofs = has_settings ? settings->get_shadow_offset() : theme_cache.font_shadow_offset;
 			int paragraph_spacing = has_settings ? settings->get_paragraph_spacing() : theme_cache.paragraph_spacing;
 			Color font_outline_color = has_settings ? settings->get_outline_color() : theme_cache.font_outline_color;
 			int outline_size = has_settings ? settings->get_outline_size() : theme_cache.font_outline_size;
@@ -875,7 +820,7 @@ void Label::_notification(int p_what) {
 
 						// Draw shadow outline.
 						if (font_shadow_color.a != 0 && shadow_outline_size > 0) {
-							DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, font_shadow_color, draw_glyph_shadow_outline, shadow_outline_size, shadow_ofs);
+							DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, font_shadow_color, draw_glyph_shadow_outline, static_cast<int>(shadow_outline_size), shadow_ofs);
 						}
 
 						// Draw shadow.
@@ -890,10 +835,11 @@ void Label::_notification(int p_what) {
 							for (int draw_iteration_index = draw_iterations - 1; draw_iteration_index >= 0; --draw_iteration_index) {
 								LabelSettings::StackedShadowData stacked_shadow_data = stacked_shadow_datas[draw_iteration_index];
 								if (stacked_shadow_data.outline_size > 0) {
-									DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, stacked_shadow_data.color, draw_glyph_shadow_outline, stacked_shadow_data.outline_size, stacked_shadow_data.offset);
+									const Vector2 shadow_ofs = stacked_shadow_data.offset;
+									DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, stacked_shadow_data.color, draw_glyph_shadow_outline, static_cast<int>(stacked_shadow_data.outline_size), shadow_ofs);
 								}
-
-								DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, stacked_shadow_data.color, draw_glyph_shadow, stacked_shadow_data.offset);
+								const Vector2 shadow_ofs = stacked_shadow_data.offset;
+								DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, stacked_shadow_data.color, draw_glyph_shadow, shadow_ofs);
 							}
 						}
 
@@ -916,14 +862,14 @@ void Label::_notification(int p_what) {
 								if (stacked_outline_data.size <= 0) {
 									continue;
 								}
-								DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, stacked_outline_data.color, draw_glyph_outline, stacked_outline_draw_size);
+								DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, stacked_outline_data.color, draw_glyph_outline, static_cast<int>(stacked_outline_draw_size));
 								stacked_outline_draw_size -= stacked_outline_data.size;
 							}
 						}
 
 						// Draw outline.
 						if (outline_size > 0 && font_outline_color.a != 0) {
-							DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, font_outline_color, draw_glyph_outline, outline_size);
+							DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, font_outline_color, draw_glyph_outline, static_cast<int>(outline_size));
 						}
 
 						// Draw text.
diff --git a/scene/gui/label.h b/scene/gui/label.h
index 1a637c2330..df608526ea 100644
--- a/scene/gui/label.h
+++ b/scene/gui/label.h
@@ -193,3 +193,56 @@ public:
 	Label(const String &p_text = String());
 	~Label();
 };
+
+template <typename... VarArgs>
+inline void DRAW_TEXT(bool rtl, int ellipsis_pos, int ellipsis_gl_size, const Glyph *ellipsis_glyphs, bool trim_chars, int para_start, int visible_chars, bool trim_glyphs_ltr, int &processed_glyphs_step, int processed_glyphs, int visible_glyphs, bool trim_glyphs_rtl, int total_glyphs, const RID &ci, const Vector2 &ofs, int gl_size, int trim_pos, const Glyph *glyphs, const Color &color, void (*draw_func)(const Glyph &p_gl, const RID &p_canvas, const Color &p_font_outline_color, const Vector2 &p_ofs, VarArgs... p_args), VarArgs &&...p_args) {
+	do {
+		processed_glyphs_step = processed_glyphs;
+		Vector2 offset_step = ofs; /* Draw RTL ellipsis string when necessary. */
+		if (rtl && ellipsis_pos >= 0) {
+			for (int gl_idx = ellipsis_gl_size - 1; gl_idx >= 0; gl_idx--) {
+				for (int j = 0; j < ellipsis_glyphs[gl_idx].repeat; j++) {
+					bool skip = (trim_chars && ellipsis_glyphs[gl_idx].end + para_start > visible_chars) || (trim_glyphs_ltr && (processed_glyphs_step >= visible_glyphs)) || (trim_glyphs_rtl && (processed_glyphs_step < total_glyphs - visible_glyphs));
+					if (!skip) {
+						draw_func(ellipsis_glyphs[gl_idx], ci, color, offset_step, p_args...);
+					}
+					processed_glyphs_step++;
+					offset_step.x += ellipsis_glyphs[gl_idx].advance;
+				}
+			}
+		} /* Draw main text. */
+		for (int j = 0; j < gl_size; j++) { /* Trim when necessary. */
+			if (trim_pos >= 0) {
+				if (rtl) {
+					if (j < trim_pos) {
+						continue;
+					}
+				} else {
+					if (j >= trim_pos) {
+						break;
+					}
+				}
+			}
+			for (int k = 0; k < glyphs[j].repeat; k++) {
+				bool skip = (trim_chars && glyphs[j].end + para_start > visible_chars) || (trim_glyphs_ltr && (processed_glyphs_step >= visible_glyphs)) || (trim_glyphs_rtl && (processed_glyphs_step < total_glyphs - visible_glyphs));
+				if (!skip) {
+					draw_func(glyphs[j], ci, color, offset_step, p_args...);
+				}
+				processed_glyphs_step++;
+				offset_step.x += glyphs[j].advance;
+			}
+		} /* Draw LTR ellipsis string when necessary. */
+		if (!rtl && ellipsis_pos >= 0) {
+			for (int gl_idx = 0; gl_idx < ellipsis_gl_size; gl_idx++) {
+				for (int j = 0; j < ellipsis_glyphs[gl_idx].repeat; j++) {
+					bool skip = (trim_chars && ellipsis_glyphs[gl_idx].end + para_start > visible_chars) || (trim_glyphs_ltr && (processed_glyphs_step >= visible_glyphs)) || (trim_glyphs_rtl && (processed_glyphs_step < total_glyphs - visible_glyphs));
+					if (!skip) {
+						draw_func(ellipsis_glyphs[gl_idx], ci, color, offset_step, p_args...);
+					}
+					processed_glyphs_step++;
+					offset_step.x += ellipsis_glyphs[gl_idx].advance;
+				}
+			}
+		}
+	} while (false);
+}

```

Delsin-Yu commented:
> Here is a patch that addresses [#104731 (comment)]
> ... Patch Content ...

I'm having trouble applying the patch content. Can you modify the patch to be based on [a7f6ab038a552442afd52e3e4a4a83845ab0d5f3]? Sorry for the inconvenience.

beicause commented:
>I'm having trouble applying the patch content.

What happened? Applying the patch on https://github.com/godotengine/godot/pull/104731/commits/a7f6ab038a552442afd52e3e4a4a83845ab0d5f3 works for me.

beicause commented:
I found a better way that can just replace the macro after I learned more about cpp template. [draw_text.zip](https://github.com/user-attachments/files/19808668/draw_text.zip)

 When applying patch, make sure there are no other local changes that could cause conflicts. Or you can copy this directly.

<details>

```diff
diff --git a/scene/gui/label.cpp b/scene/gui/label.cpp
index 4387577e58..ad92af33b1 100644
--- a/scene/gui/label.cpp
+++ b/scene/gui/label.cpp
@@ -445,61 +445,6 @@ inline void draw_glyph_outline(const Glyph &p_gl, const RID &p_canvas, const Col
 	}
 }
 
-#define DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para_start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, color, draw_func, ...) \
-	do {                                                                                                                                                                                                                                                                           \
-		processed_glyphs_step = processed_glyphs;                                                                                                                                                                                                                                  \
-		Vector2 offset_step = ofs;                                                                                                                                                                                                                                                 \
-		/* Draw RTL ellipsis string when necessary. */                                                                                                                                                                                                                             \
-		if (rtl && ellipsis_pos >= 0) {                                                                                                                                                                                                                                            \
-			for (int gl_idx = ellipsis_gl_size - 1; gl_idx >= 0; gl_idx--) {                                                                                                                                                                                                       \
-				for (int j = 0; j < ellipsis_glyphs[gl_idx].repeat; j++) {                                                                                                                                                                                                         \
-					bool skip = (trim_chars && ellipsis_glyphs[gl_idx].end + para_start > visible_chars) || (trim_glyphs_ltr && (processed_glyphs_step >= visible_glyphs)) || (trim_glyphs_rtl && (processed_glyphs_step < total_glyphs - visible_glyphs));                        \
-					if (!skip) {                                                                                                                                                                                                                                                   \
-						draw_func(ellipsis_glyphs[gl_idx], ci, color, offset_step, ##__VA_ARGS__);                                                                                                                                                                                 \
-					}                                                                                                                                                                                                                                                              \
-					processed_glyphs_step++;                                                                                                                                                                                                                                       \
-					offset_step.x += ellipsis_glyphs[gl_idx].advance;                                                                                                                                                                                                              \
-				}                                                                                                                                                                                                                                                                  \
-			}                                                                                                                                                                                                                                                                      \
-		}                                                                                                                                                                                                                                                                          \
-		/* Draw main text. */                                                                                                                                                                                                                                                      \
-		for (int j = 0; j < gl_size; j++) {                                                                                                                                                                                                                                        \
-			/* Trim when necessary. */                                                                                                                                                                                                                                             \
-			if (trim_pos >= 0) {                                                                                                                                                                                                                                                   \
-				if (rtl) {                                                                                                                                                                                                                                                         \
-					if (j < trim_pos) {                                                                                                                                                                                                                                            \
-						continue;                                                                                                                                                                                                                                                  \
-					}                                                                                                                                                                                                                                                              \
-				} else {                                                                                                                                                                                                                                                           \
-					if (j >= trim_pos) {                                                                                                                                                                                                                                           \
-						break;                                                                                                                                                                                                                                                     \
-					}                                                                                                                                                                                                                                                              \
-				}                                                                                                                                                                                                                                                                  \
-			}                                                                                                                                                                                                                                                                      \
-			for (int k = 0; k < glyphs[j].repeat; k++) {                                                                                                                                                                                                                           \
-				bool skip = (trim_chars && glyphs[j].end + para_start > visible_chars) || (trim_glyphs_ltr && (processed_glyphs_step >= visible_glyphs)) || (trim_glyphs_rtl && (processed_glyphs_step < total_glyphs - visible_glyphs));                                          \
-				if (!skip) {                                                                                                                                                                                                                                                       \
-					draw_func(glyphs[j], ci, color, offset_step, ##__VA_ARGS__);                                                                                                                                                                                                   \
-				}                                                                                                                                                                                                                                                                  \
-				processed_glyphs_step++;                                                                                                                                                                                                                                           \
-				offset_step.x += glyphs[j].advance;                                                                                                                                                                                                                                \
-			}                                                                                                                                                                                                                                                                      \
-		}                                                                                                                                                                                                                                                                          \
-		/* Draw LTR ellipsis string when necessary. */                                                                                                                                                                                                                             \
-		if (!rtl && ellipsis_pos >= 0) {                                                                                                                                                                                                                                           \
-			for (int gl_idx = 0; gl_idx < ellipsis_gl_size; gl_idx++) {                                                                                                                                                                                                            \
-				for (int j = 0; j < ellipsis_glyphs[gl_idx].repeat; j++) {                                                                                                                                                                                                         \
-					bool skip = (trim_chars && ellipsis_glyphs[gl_idx].end + para_start > visible_chars) || (trim_glyphs_ltr && (processed_glyphs_step >= visible_glyphs)) || (trim_glyphs_rtl && (processed_glyphs_step < total_glyphs - visible_glyphs));                        \
-					if (!skip) {                                                                                                                                                                                                                                                   \
-						draw_func(ellipsis_glyphs[gl_idx], ci, color, offset_step, ##__VA_ARGS__);                                                                                                                                                                                 \
-					}                                                                                                                                                                                                                                                              \
-					processed_glyphs_step++;                                                                                                                                                                                                                                       \
-					offset_step.x += ellipsis_glyphs[gl_idx].advance;                                                                                                                                                                                                              \
-				}                                                                                                                                                                                                                                                                  \
-			}                                                                                                                                                                                                                                                                      \
-		}                                                                                                                                                                                                                                                                          \
-	} while (false)
-
 void Label::_ensure_shaped() const {
 	if (dirty || font_dirty || text_dirty) {
 		_shape();
@@ -875,12 +820,12 @@ void Label::_notification(int p_what) {
 
 						// Draw shadow outline.
 						if (font_shadow_color.a != 0 && shadow_outline_size > 0) {
-							DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, font_shadow_color, draw_glyph_shadow_outline, shadow_outline_size, shadow_ofs);
+							draw_text(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, font_shadow_color, draw_glyph_shadow_outline, shadow_outline_size, shadow_ofs);
 						}
 
 						// Draw shadow.
 						if (font_shadow_color.a > 0) {
-							DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, font_shadow_color, draw_glyph_shadow, shadow_ofs);
+							draw_text(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, font_shadow_color, draw_glyph_shadow, shadow_ofs);
 						}
 
 						// Draw stacked shadow.
@@ -890,10 +835,10 @@ void Label::_notification(int p_what) {
 							for (int draw_iteration_index = draw_iterations - 1; draw_iteration_index >= 0; --draw_iteration_index) {
 								LabelSettings::StackedShadowData stacked_shadow_data = stacked_shadow_datas[draw_iteration_index];
 								if (stacked_shadow_data.outline_size > 0) {
-									DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, stacked_shadow_data.color, draw_glyph_shadow_outline, stacked_shadow_data.outline_size, stacked_shadow_data.offset);
+									draw_text(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, stacked_shadow_data.color, draw_glyph_shadow_outline, stacked_shadow_data.outline_size, stacked_shadow_data.offset);
 								}
 
-								DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, stacked_shadow_data.color, draw_glyph_shadow, stacked_shadow_data.offset);
+								draw_text(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, stacked_shadow_data.color, draw_glyph_shadow, stacked_shadow_data.offset);
 							}
 						}
 
@@ -916,19 +861,19 @@ void Label::_notification(int p_what) {
 								if (stacked_outline_data.size <= 0) {
 									continue;
 								}
-								DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, stacked_outline_data.color, draw_glyph_outline, stacked_outline_draw_size);
+								draw_text(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, stacked_outline_data.color, draw_glyph_outline, stacked_outline_draw_size);
 								stacked_outline_draw_size -= stacked_outline_data.size;
 							}
 						}
 
 						// Draw outline.
 						if (outline_size > 0 && font_outline_color.a != 0) {
-							DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, font_outline_color, draw_glyph_outline, outline_size);
+							draw_text(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, font_outline_color, draw_glyph_outline, outline_size);
 						}
 
 						// Draw text.
 						{
-							DRAW_TEXT(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, font_color, draw_glyph);
+							draw_text(rtl, ellipsis_pos, ellipsis_gl_size, ellipsis_glyphs, trim_chars, para.start, visible_chars, trim_glyphs_ltr, processed_glyphs_step, processed_glyphs, visible_glyphs, trim_glyphs_rtl, total_glyphs, ci, ofs, gl_size, trim_pos, glyphs, font_color, draw_glyph);
 						}
 
 						processed_glyphs = processed_glyphs_step;
diff --git a/scene/gui/label.h b/scene/gui/label.h
index 1a637c2330..9c1756070f 100644
--- a/scene/gui/label.h
+++ b/scene/gui/label.h
@@ -192,4 +192,57 @@ public:
 
 	Label(const String &p_text = String());
 	~Label();
+
+	template <typename... VarArgsFunc, typename... VarArgs>
+	void draw_text(bool rtl, int ellipsis_pos, int ellipsis_gl_size, const Glyph *ellipsis_glyphs, bool trim_chars, int para_start, int visible_chars, bool trim_glyphs_ltr, int &processed_glyphs_step, int processed_glyphs, int visible_glyphs, bool trim_glyphs_rtl, int total_glyphs, const RID &ci, const Vector2 &ofs, int gl_size, int trim_pos, const Glyph *glyphs, const Color &color, void (*draw_func)(const Glyph &p_gl, const RID &p_canvas, const Color &p_font_outline_color, const Vector2 &p_ofs, VarArgsFunc... p_args), VarArgs &&...p_args) {
+		do {
+			processed_glyphs_step = processed_glyphs;
+			Vector2 offset_step = ofs; /* Draw RTL ellipsis string when necessary. */
+			if (rtl && ellipsis_pos >= 0) {
+				for (int gl_idx = ellipsis_gl_size - 1; gl_idx >= 0; gl_idx--) {
+					for (int j = 0; j < ellipsis_glyphs[gl_idx].repeat; j++) {
+						bool skip = (trim_chars && ellipsis_glyphs[gl_idx].end + para_start > visible_chars) || (trim_glyphs_ltr && (processed_glyphs_step >= visible_glyphs)) || (trim_glyphs_rtl && (processed_glyphs_step < total_glyphs - visible_glyphs));
+						if (!skip) {
+							draw_func(ellipsis_glyphs[gl_idx], ci, color, offset_step, std::forward<VarArgs>(p_args)...);
+						}
+						processed_glyphs_step++;
+						offset_step.x += ellipsis_glyphs[gl_idx].advance;
+					}
+				}
+			} /* Draw main text. */
+			for (int j = 0; j < gl_size; j++) { /* Trim when necessary. */
+				if (trim_pos >= 0) {
+					if (rtl) {
+						if (j < trim_pos) {
+							continue;
+						}
+					} else {
+						if (j >= trim_pos) {
+							break;
+						}
+					}
+				}
+				for (int k = 0; k < glyphs[j].repeat; k++) {
+					bool skip = (trim_chars && glyphs[j].end + para_start > visible_chars) || (trim_glyphs_ltr && (processed_glyphs_step >= visible_glyphs)) || (trim_glyphs_rtl && (processed_glyphs_step < total_glyphs - visible_glyphs));
+					if (!skip) {
+						draw_func(glyphs[j], ci, color, offset_step, std::forward<VarArgs>(p_args)...);
+					}
+					processed_glyphs_step++;
+					offset_step.x += glyphs[j].advance;
+				}
+			} /* Draw LTR ellipsis string when necessary. */
+			if (!rtl && ellipsis_pos >= 0) {
+				for (int gl_idx = 0; gl_idx < ellipsis_gl_size; gl_idx++) {
+					for (int j = 0; j < ellipsis_glyphs[gl_idx].repeat; j++) {
+						bool skip = (trim_chars && ellipsis_glyphs[gl_idx].end + para_start > visible_chars) || (trim_glyphs_ltr && (processed_glyphs_step >= visible_glyphs)) || (trim_glyphs_rtl && (processed_glyphs_step < total_glyphs - visible_glyphs));
+						if (!skip) {
+							draw_func(ellipsis_glyphs[gl_idx], ci, color, offset_step, std::forward<VarArgs>(p_args)...);
+						}
+						processed_glyphs_step++;
+						offset_step.x += ellipsis_glyphs[gl_idx].advance;
+					}
+				}
+			}
+		} while (false);
+	}
 };

```
</details>

Delsin-Yu commented:
> I found a better way that can just replace the macro after I learned more about cpp template.

Thanks, I ended up manually applying the patch, and I can confirm that the patch does indeed work.

KoBeWi commented on scene/resources/label_settings.cpp:
```diff
@@ -81,6 +105,23 @@ void LabelSettings::_bind_methods() {
+
+	stacked_outline_base_property_helper.set_prefix("stacked_outline_");
+	stacked_outline_base_property_helper.set_array_length_getter(&LabelSettings::get_stacked_outline_count);
+	stacked_outline_base_property_helper.register_property(PropertyInfo(Variant::INT, "size", PROPERTY_HINT_NONE, "0,127,1,or_greater,suffix:px"), 0, &LabelSettings::set_stacked_outline_size, &LabelSettings::get_stacked_outline_size);
```
Instead of hard-coding defaults, you should create an instance of `StackedOutlineData` and use its properties. See here for example:
https://github.com/godotengine/godot/blob/4e6451d62a931a664acabbf3967e2874fedaadcd/scene/gui/file_dialog.cpp#L1612-L1616

KoBeWi reviewed (approved):
LabelSetting changes and the functionality itself look fine.
Label code could use a review, as the changes look rather complex.

Calinou reviewed (approved):
Tested again, it works as expected. I like the new UI more, as the options are now more visually connected to each other in the inspector. The code changes are also less intrusive.

**Updated testing project:** [test_pr_104731.zip](https://github.com/user-attachments/files/19914602/test_pr_104731.zip)

Mickeon commented on doc/classes/LabelSettings.xml:
```diff
@@ -8,6 +8,127 @@
+			<return type="void" />
+			<param index="0" name="index" type="int" default="-1" />
+			<description>
+				Adds a new stacked outline to the label at the given [param index]. If [param index] is -1, the new stacked outline will be added at the end of the list.
```
```suggestion
				Adds a new stacked outline to the label at the given [param index]. If [param index] is [code]-1[/code], the new stacked outline will be added at the end of the list.
```
What is this "-1" behaviour inspired by? If it's meant to be like Array, negative indices should be supported. Otherwise, I'm not sure? Is it like... TreeItem, PopupMenu?

Delsin-Yu replied:
I think me at that time is based on the design in `TileSet`
![image](https://github.com/user-attachments/assets/150486e7-a2f1-4ea4-9c63-0d8c49aded08)

Mickeon replied:
Alright, seems about right.
Although, it's a lie to mention `-1` as the condition actually checks for negatives, but... whatever. Maybe it makes the purpose of the default value clearer.

Nothing against it, I just wanted to make sure there was consistency.

Mickeon commented on doc/classes/LabelSettings.xml:
```diff
@@ -8,6 +8,127 @@
+			<return type="void" />
+			<param index="0" name="index" type="int" default="-1" />
+			<description>
+				Adds a new stacked shadow to the label at the given [param index]. If [param index] is -1, the new stacked shadow will be added at the end of the list.
```
```suggestion
				Adds a new stacked shadow to the label at the given [param index]. If [param index] is [code]-1[/code], the new stacked shadow will be added at the end of the list.
```

Mickeon commented on doc/classes/LabelSettings.xml:
```diff
@@ -8,6 +8,127 @@
+			<param index="0" name="index" type="int" />
+			<param index="1" name="color" type="Color" />
+			<description>
+				Sets the color of the stacked outline identified by the given index.
```
The existing description is fine with context, but I personally feel like it's incomplete because it omits the `color` parameter.
```suggestion
				Sets the color of the stacked outline identified by the given [param index] to [param color].
```
Maybe it's okay, though?
A similar suggestion would also apply also to the remaining methods below.

Mickeon commented on doc/classes/LabelSettings.xml:
```diff
@@ -39,5 +160,11 @@
 			Size of the shadow effect.
 		</member>
+		<member name="stacked_outline_count" type="int" setter="set_stacked_outline_count" getter="get_stacked_outline_count" default="0">
+			Returns the stacked outline count.
```
This is not a method, it's a property. At the very least:
```suggestion
			The number of stacked outlines.
```
But it could be elaborated on a bit more (what happens when this value changes, and so on).
Which makes me think, should it even be exposed this way? Why is the setter exposed? I assume it's for editor reasons.

Delsin-Yu replied:
I think the setter is required by `ADD_ARRAY_COUNT`, but I'm new to this macro

![image](https://github.com/user-attachments/assets/9154e4b6-a9ff-446d-a968-33d105b6e73b)

Mickeon replied:
Without the setter being exposed I seriously doubt the macro, or the property itself, would even work.
It's fine, then.

Mickeon reviewed (approved):
Docs are okay

Repiteo commented:
Thanks!
